tiledb/api/src/array/enumeration/mod.rs

```diff
@@ -0,0 +1,529 @@
+        // Some and None. Here we're just returning None directly to avoid
+        // the issue.
+        if ptr.is_null() {
+            return Ok(None);
+        }
+
+        // The size returned is the number of bytes, where from_raw_parts
+        // wants the number of elements.
+        assert!(size as usize % std::mem::size_of::<u64>() == 0);
+        let elems = size as usize / std::mem::size_of::<u64>();
+        let slice = unsafe { std::slice::from_raw_parts(ptr, elems) };
+        Ok(Some(slice))
+    }
+
+    pub fn extend<T>(
+        &self,
+        data: &[T],
+        offsets: Option<&[u64]>,
+    ) -> TileDBResult<Enumeration<'ctx>> {
+        let mut c_new_enmr: *mut ffi::tiledb_enumeration_t = out_ptr!();
+
+        // Rust semantics require that slice pointers aren't nullptr so that
+        // nullptr can be used to distinguish between Some and None. The stdlib
+        // empty slices all appear to return 0x1 which is mentioned in the docs
+        // as a valid strategy.
+        let (offsets_ptr, offsets_len) = if offsets.is_none() {
+            (std::ptr::null_mut() as *const u64, 0u64)
+        } else {
+            let offsets = offsets.unwrap();
+            (offsets.as_ptr(), std::mem::size_of_val(offsets) as u64)
+        };
+
+        // An important note here is that the Enumeration allocator copies the
+        // contents of data of offsets rather than assumes ownership. That
+        // means this is safe as those bytes are guaranteed to be alive until
+        // we drop self at the end of this method after returning from
+        // tiledb_enumeration_alloc.
+        let res = unsafe {
+            ffi::tiledb_enumeration_extend(
+                self.context.capi(),
+                self.capi(),
+                data.as_ptr() as *const std::ffi::c_void,
+                std::mem::size_of_val(data) as u64,
+                offsets_ptr as *const std::ffi::c_void,
+                offsets_len,
+                &mut c_new_enmr,
+            )
+        };
+
+        if res == ffi::TILEDB_OK {
+            Ok(Enumeration {
+                context: self.context,
+                raw: RawEnumeration::Owned(c_new_enmr),
+            })
+        } else {
+            Err(self.context.expect_last_error())
+        }
+    }
+}
+
+impl<'ctx> Debug for Enumeration<'ctx> {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        let name = self.name().map_err(|_| fmt::Error)?;
+        let dtype = self.datatype().map_err(|_| fmt::Error)?;
+
+        let dtype_string =
+            dtype.to_string().unwrap_or("<unknown datatype>".to_owned());
+        let cell_val_num = self.cell_val_num().map_err(|_| fmt::Error)?;
+        let ordered = self.ordered().map_err(|_| fmt::Error)?;
+
+        let json = json!({
+            "name": name,
+            "datatype": dtype_string,
+            "cell_val_num": cell_val_num,
+            "ordered": ordered,
+            "values": [], // TODO: Render values
```

Comment:
Other impls of Debug construct the corresponding `ThingData` and then format that. That would get you the value formatting here.

Reply:
The part I was being lazy on is the conversion of `Box<[u8]>` back into the datatype it came in as. Now that I have the basics figured out after fighting with the whole "fn_typed calls a generic function" I'll spend some extra time writing some conversion utilities and then some strategies for using those to generate data based on a datatype/cell_val_num combination. Only question is if you want me to do that as part of this PR or separate it out?

Reply:
I think we can do a separate PR

Reply:
Will do.